Read the bearer token over websocket endpoints instead of query parameter

lxea opened the pull request:
Original PR here: gravitational/teleport-private#1201

Fixes gravitational/teleport-private#637

Changelog: Read bearer token via websocket on websocket endpoints

Commit:
Read the bearer token over WS endpoints

use the request context, not session

Dont pass websocket by context

lint

resolve some comments

Add TestWSAuthenticateRequest

Close ws in handler

deprecation notices, doc

resolve comments

resolve comments

give a longer read/write deadline

dont set write deadline, ws endpoints never did before and it breaks things

convert frontend to use ws access token

Resolove comments, move to using an explicit state

fix ci

reset read deadline

prettier

Commit:
update connectToHost

Commit:
linter

rosstimothy reviewed (commented):
It looks like the changes made here have broken `TestSiteNodeConnectInvalidSessionID`

lxea commented:
> It looks like the changes made here have broken `TestSiteNodeConnectInvalidSessionID`

Yeah, trying to debug it at the moment

lxea commented on lib/web/apiserver.go:
```diff
@@ -4118,6 +4208,68 @@ func contextWithMFAResponseFromRequestHeader(ctx context.Context, requestHeader
+	Message string `json:"message,omitempty"`
+}
+
+var wsIODeadline = time.Second * 4
```
it does get over written in a test

Commit:
read errors from websocket

zmb3 commented:
Desktop access looks good, but I can't start SSH sessions from the web with this branch.

<img width="1563" alt="image" src="https://github.com/gravitational/teleport/assets/7527103/c7cf1687-cede-4330-bd47-f5037f396ac6">

Commit:
missing /ws on ttyWsAddr and fix wrong onmessage

lxea commented:
> Desktop access looks good, but I can't start SSH sessions from the web with this branch.
> <img alt="image" width="1563" src="https://private-user-images.githubusercontent.com/7527103/300967039-c7cf1687-cede-4330-bd47-f5037f396ac6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2OTczMDYsIm5iZiI6MTcwNjY5NzAwNiwicGF0aCI6Ii83NTI3MTAzLzMwMDk2NzAzOS1jN2NmMTY4Ny1jZWRlLTQzMzAtYmQ0Ny1mNTAzN2YzOTZhYzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEzMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMzFUMTAzMDA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzEwNjM2MjRkNTk0NzYyZGMzNGM5ZjRlMmYzM2ZhZGQzZmEzMGI1ZTViZTRlNTU0NDlhMTQyNTkwMjczMDFhZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.PEu3X3Pm91BkMoAJReMV5NfW46PCv9caC8lJzZpU8FE">

should be working now

Commit:
fix race in test

Commit:
lint

Commit:
skip TestTerminal as it takes 11 seconds to run

lxea commented on build.assets/tooling/cmd/difftest/main.go:
```diff
@@ -76,6 +76,9 @@ var (
 
 		// TestAdminActionMFA takes longer than 6 seconds to run.
 		"TestAdminActionMFA",
+
+		// TestTerminal takes 11 seconds to run.
+		"TestTerminal",
```
What do you mean? Like force the pr to merge?

lxea replied:
Would that not cause problems for future prs that run the TestTerminal test in the flaky check?

Commit:
dont skip the test

zmb3 commented:
Just confirming that I tested again this morning and the SSH issue is fixed.

ibeckermayer commented on lib/web/apiserver.go:
```diff
@@ -4118,6 +4239,68 @@ func contextWithMFAResponseFromRequestHeader(ctx context.Context, requestHeader
+
+// AuthenticateRequest authenticates request using combination of a session cookie
+// and bearer token retrieved from a websocket
+func (h *Handler) AuthenticateRequestWS(w http.ResponseWriter, r *http.Request) (*SessionContext, *websocket.Conn, error) {
```
I notice that the corollary `AuthenticateRequest` method has these lines:

https://github.com/gravitational/teleport/blob/f8ad58c7c0194029aba1ea3ad1187dc382558f5f/lib/web/apiserver.go#L4200-L4202

which I don't see replicated anywhere in here. That said, I tested this with mfa required and everything worked as planned. Do you know why that's not necessary in this path?

lxea replied:
I dont know why it works when not included, these lines were added after i started the PR so I never noticed it, i think it makes sense to include them

ibeckermayer commented on lib/web/apiserver.go:
```diff
@@ -4118,6 +4239,68 @@ func contextWithMFAResponseFromRequestHeader(ctx context.Context, requestHeader
+	Message string `json:"message,omitempty"`
+}
+
+var wsIODeadline = time.Second * 4
```
A comment explaining what this means would be helpful for future devs.

ibeckermayer commented on lib/web/apiserver.go:
```diff
@@ -4118,6 +4239,68 @@ func contextWithMFAResponseFromRequestHeader(ctx context.Context, requestHeader
+	if err != nil {
+		return nil, nil, trace.ConnectionProblem(err, "Error upgrading to websocket: %v", err)
+	}
+	if err := ws.SetReadDeadline(deadlineForInterval(wsIODeadline)); err != nil {
```
It doesn't quite make sense to pass something called `wsIODeadline` into a function named `deadlineForInterval`. `deadlineForInterval` is built for taking in an "interval" and returning a "deadline". I don't know that "interval" really makes sense in this context, seems like simplest would just be

```suggestion
	if err := ws.SetReadDeadline(time.Now().Add(wsIODeadline)); err != nil {
```

(note that `deadlineForInterval` was multiplying `wsIODeadline` by 2, so you may want to adjust it if you go with this suggestion).

ibeckermayer commented on lib/web/apiserver.go:
```diff
@@ -4118,6 +4239,68 @@ func contextWithMFAResponseFromRequestHeader(ctx context.Context, requestHeader
+		return nil, nil, trace.Wrap(err)
+	}
+
+	if err := ws.SetReadDeadline(time.Time{}); err != nil {
```
A comment noting that you're unsetting the deadline to allow the connection to proceed however it likes could be helpful.

ibeckermayer commented on lib/web/apiserver.go:
```diff
@@ -4118,6 +4239,68 @@ func contextWithMFAResponseFromRequestHeader(ctx context.Context, requestHeader
+	}
+
+	if err := ws.SetReadDeadline(time.Time{}); err != nil {
+		log.WithError(err).Error("Error setting websocket read deadline")
```
nit: is there a particular reason that errors have to be logged when `SetReadDeadline` fails (here and above) but not for other errors?

ibeckermayer commented on lib/web/apiserver_test.go:
```diff
@@ -9066,6 +9102,109 @@ func (s *fakeKubeService) ListKubernetesResources(ctx context.Context, req *kube
+			serverExpectError: "i/o timeout",
+			token:             pack.session.Token,
+			readTimeout: func() {
+				<-time.After(time.Second * 3)
```
```suggestion
				<-time.After(wsIODeadline * 3)
```

ibeckermayer commented on web/packages/teleport/src/Assist/context/AssistContext.tsx:
```diff
@@ -448,15 +471,41 @@ export function AssistContextProvider(props: PropsWithChildren<unknown>) {
     executeCommandWebSocket.current = new WebSocket(url);
-    executeCommandWebSocket.current.binaryType = 'arraybuffer';
 
+    executeCommandWebSocket.current.onopen = () => {
+      executeCommandWebSocket.current.send(
+        JSON.stringify({ token: getAccessToken() })
+      );
+    };
```
Seems to me that we have a a similar sort of thing in several places, wherein we have a `websocket.onopen` which needs to send the json-ified access token, and then parse the expected response in a certain way `onmessage` before flipping an `authenticated` flag. This makes me think that we might want to provide a shared `AuthenticatingWebSocket` that wraps a native `WebSocket` object and keeps all this logic in one place, and then swap these various individual `WebSocket`s out for that.

We'll have to get the API right so that you can basically use it like an ordinary `WebSocket`, but with our little authentication protocol automatically sprinkled on top; It might take a bit of work, but it will be a good way to keep the implementation of authenticating websockets consistent throughout the codebase going forward. WDYT?

ryanclark replied:
This would be a great solution, especially as it seems we need some logic to wait for an authenticated connection before attempting to send any messages

lxea replied:
I'm not particularly confident when it comes to good javascript interface design, but I've pushed a version with something like this and a version working when connecting to a node, lmk what suggestions ye might have.

lxea commented on lib/web/apiserver.go:
```diff
@@ -3726,12 +3747,103 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
+		upgrader := websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+			CheckOrigin:     func(r *http.Request) bool { return true },
```
What would be preferred here, I dont fully understand cors, would it be set like this so connections can be made to other proxies on different clusters?

ibeckermayer replied:
https://github.com/gravitational/teleport/pull/37520/commits/39c56a4197b26402f42081c513868752fa076681

I hunted this down to an old comment [here](https://github.com/gravitational/teleport/blob/84b64fe487f9a74c64f702b65e2dc7bca3eb42a5/lib/web/terminal.go#L192-L198), since we're maintaining the old behavior then we'll consider it out of scope for this PR. I consolidated the question into a single global upgrader in https://github.com/gravitational/teleport/pull/37520/commits/39c56a4197b26402f42081c513868752fa076681.

Commit:
resolve apiserver comments

Commit:
Add an AuthenticatedWebSocket class

Commit:
convert other clients to use AuthenticatedWebSocket

Commit:
Converts `AuthenticatedWebSocket` into drop-in replacement for `WebSocket` (#37699)

* Converts `AuthenticatedWebSocket` into drop-in replacement for `WebSocket`
that automatically goes through Teleport's custom authentication process
before facilitating any caller-defined communication.

This also reverts previous-`WebSocket` users to their original state
(sans the code for passing the bearer token in the query string),
swapping in `AuthenticatedWebSocket` in place of `WebSocket`.

Commit:
Create a single authnWsUpgrader with a comment justifying why we turn off CORS

Commit:
recieving to receiving

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -3731,12 +3752,102 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
 	})
 }
 
+func (h *Handler) writeErrToWS(ws *websocket.Conn, err error) {
```
```suggestion
func (h *Handler) writeErrToWebsocket(ws *websocket.Conn, err error) {
```

Let's spell out websocket - it's not that much to type and `WS` could also be taken to mean `WebSession`.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -3731,12 +3752,102 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
+func (h *Handler) writeErrToWS(ws *websocket.Conn, err error) {
+	errEnvelope := Envelope{
+		Type:    defaults.WebsocketError,
+		Payload: err.Error(),
```
We should try an extract `trace.UserMessage` first. Sending an arbitrary error all the way to the frontend is likely to result in poor UX or even divulging info that shouldn't be.

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -3731,12 +3752,102 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
+// as specified by the ":site" url parameter.
+//
+// TODO(lxea): remove the 'websocketAuth' bool once the deprecated websocket handlers are removed
+func (h *Handler) WithClusterAuthWS(websocketAuth bool, fn ClusterWebsocketHandler) httprouter.Handle {
```
```suggestion
func (h *Handler) WithClusterAuthWebsocket(websocketAuth bool, fn ClusterWebsocketHandler) httprouter.Handle {
```

zmb3 commented on lib/web/apiserver_test.go:
```diff
@@ -8117,18 +8129,38 @@ func (r *testProxy) newClient(t *testing.T, opts ...roundtrip.ClientParam) *Test
+		return trace.Wrap(err)
+	}
+
+	if err := ws.WriteMessage(websocket.TextMessage, authReq); err != nil {
```
Do any of our clients care about the switch from `TextMessage` to `BinaryMessage`?

lxea replied:
Not that I can tell, it doesnt seem to be explicitly taken into account at least

r0mant reviewed (approved):
@jakule confirmed Assist works with these changes. @lxea - let's address Zac's latest comments and then merge and backport to v15 so we can get this released.

jakule reviewed (commented):
I tested Assist and it seems to work. I also added a few comments.

r0mant commented:
/excludeflake *

lxea commented on lib/web/apiserver.go:
```diff
@@ -3731,12 +3752,102 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
+				return nil, trace.Wrap(err)
+			}
+
+			defer ws.Close()
```
Left a note saying downstream users should be sending the close

Commit:
resolve comments

Commit:
Merge branch 'master' into lxea/websocket-bearer-token
